src/ui/Button.cpp

```diff
@@ -0,0 +1,204 @@
+﻿// Button.cpp
+
+#include "../main.hpp"
+#include "../player.hpp"
+#include "../draw.hpp"
+#include "Frame.hpp"
+#include "Button.hpp"
+#include "Image.hpp"
+#include "Text.hpp"
+
+Button::Button() {
+	size.x = 0; size.w = 32;
+	size.y = 0; size.h = 32;
+	color = SDL_MapRGBA(mainsurface->format, 127, 127, 127, 255);
+	textColor = SDL_MapRGBA(mainsurface->format, 255, 255, 255, 255);
+}
+
+Button::Button(Frame& _parent) : Button() {
+	parent = &_parent;
+	_parent.getButtons().push_back(this);
+	_parent.adoptWidget(*this);
+}
+
+Button::~Button() {
+	if (callback) {
+		delete callback;
+		callback = nullptr;
+	}
+}
+
+void Button::setIcon(const char* _icon) {
+	icon = _icon;
+}
+
+void Button::activate() {
+	if (style == STYLE_NORMAL) {
+		setPressed(true);
+	} else {
+		setPressed(isPressed()==false);
+	}
+	Widget::Args args(params);
+	if (callback) {
+		(*callback)(args);
+	} else {
+		printlog("button clicked with no callback (script or otherwise)");
+	}
+}
+
+void Button::draw(SDL_Rect _size, SDL_Rect _actualSize) {
+	_size.x += std::max(0, size.x - _actualSize.x);
+	_size.y += std::max(0, size.y - _actualSize.y);
+	_size.w = std::min(size.w, _size.w - size.x + _actualSize.x) + std::min(0, size.x - _actualSize.x);
+	_size.h = std::min(size.h, _size.h - size.y + _actualSize.y) + std::min(0, size.y - _actualSize.y);
+	if (_size.w <= 0 || _size.h <= 0)
+		return;
+
+	if (pressed) {
+		drawDepressed(_size.x, _size.y, _size.x + _size.w, _size.y + _size.h);
+	} else {
+		drawWindow(_size.x, _size.y, _size.x + _size.w, _size.y + _size.h);
+	}
+
+	if (!text.empty() && style != STYLE_CHECKBOX) {
+		Text* _text = Text::get(text.c_str(), font.c_str());
+		if (_text) {
+			SDL_Rect pos;
+			int textX = style == STYLE_DROPDOWN ? 5 + border : _size.w / 2 - _text->getWidth() / 2;
```

Comment:
Minor suggestion: I think adding parentheses around the condition would help a bit with readability.

Reply:
sure I can break it up a bit too and improve readability.